metric_learn/rca.py

```diff
@@ -62,32 +79,42 @@ def fit(self, data, chunks):
         when ``chunks[i] == j``, point i belongs to chunklet j.
     """
     data, chunks, num_chunks, d = self._process_inputs(data, chunks)
-
-    # mean center
-    data -= data.mean(axis=0)
-
-    # mean center each chunklet separately
-    chunk_mask = chunks != -1
-    chunk_data = data[chunk_mask]
-    chunk_labels = chunks[chunk_mask]
-    for c in xrange(num_chunks):
-      mask = chunk_labels == c
-      chunk_data[mask] -= chunk_data[mask].mean(axis=0)
-
-    # "inner" covariance of chunk deviations
+    chunk_mask, chunk_data, chunk_labels = _process_chunks(data, chunks, num_chunks)
     inner_cov = np.cov(chunk_data, rowvar=0, bias=1)
+    rank = np.linalg.matrix_rank(inner_cov)
+
+    # If the inner covariance matrix is not full rank,
+    # the input data are first projected with a PCA to a space of dimension rank.
+    M_pca = np.identity(d)
+    if rank < d:
+      pca = decomposition.PCA(n_components = rank)
+      pca.fit(data)
+      M_pca = pca.components_
+      data = pca.transform(data)
+      chunk_mask, chunk_data, chunk_labels = _process_chunks(data, chunks, num_chunks)
+      inner_cov = np.cov(chunk_data, rowvar=0, bias=1)
+      rank = np.linalg.matrix_rank(inner_cov)
+
+    # The embedding dimension must be smaller than the rank of the inner covariance matrix
+    dim = min(self.params['dim'], rank)
+    if dim < self.params['dim']:
+      mess  = 'The embedding dimension must be smaller than the rank of the inner covariance matrix. '
+      mess += 'dim is set to ' + str(dim) + '.'
+      warnings.warn(mess)
 
     # Fisher Linear Discriminant projection
-    if self.params['dim'] < d:
+    if dim < rank:
       total_cov = np.cov(data[chunk_mask], rowvar=0)
       tmp = np.linalg.lstsq(total_cov, inner_cov)[0]
       vals, vecs = np.linalg.eig(tmp)
-      inds = np.argsort(vals)[:self.params['dim']]
+      inds = np.argsort(vals)[:dim]
       A = vecs[:,inds]
       inner_cov = A.T.dot(inner_cov).dot(A)
       self._transformer = _inv_sqrtm(inner_cov).dot(A.T)
     else:
       self._transformer = _inv_sqrtm(inner_cov).T
+        
+    self._transformer = self._transformer.dot(M_pca)
```

Comment:
I'd rather have `M_pca` default to `None` instead of the identity matrix, then for this line do:
```python
if M_pca is not None:
  self._transformer = self._transformer.dot(M_pca)
```